apps/design/design.md

```diff
@@ -0,0 +1,159 @@
+almost any action to and from an external cache. 
+
+It is not clear to a user that a confidential client should be per user to prevent scaling
+problems. 
+
+We cannot change the MSAL cache design at this time, therefore it should be clear that
+confidential.Client should be done per user. This must go beyond a simple doc entry
+that can be ignored. Its great to say: "we told you in the doc", but that is AFTER a support call.
+
+To make it clear, we require the user to pass a userID field. This forces the user to read what
+a userID is and recognize that it is per user.
+
+A future design might include a parent client to Confidential that keeps track of the "users".
+This way we manage the lifecycle of the clients. The current design forces the user to deal with
+this, probably via a map. We could design an implementation with features from an LRU (to avoid
+memory growth), deletion schemantics, etc.... that is more efficient than what a normal user is
+going to implement (and eliminate each user from having to do this).
+
+or.....
+
+We could create a new caching mechanism for Confidential that doesn't have this problem.  Treat
+it differently than Public. This would be a larger migration path across all SDKs, but it is
+probably the right long term move.
+
+### Use of x509.Certificate and CertFromPEM() function
+
+The original version of this package used an thumbprint and a private key to do authorizations
+based on a certificate. But there wasn't a real way to get a thumbprint.
+
+A thumbprint is defined in the Oauth spec, which we had to track down. It is an SHA-1 hash
+from the x509 certificate's DER encdoed ASN1 bytes. 
+
+Since the user was going to need the x509, we moved to having the user provide the x509.Certificate
+object. 
+
+We wrote the thumbprint creator for the internals. 
+
+Since we also require the private key and it is not straightforward to get, we added a CertFromPEM()
+function that will extract the x509.Certificate and private key as long as PEM is PKCS8 encoded. We
+probably could add PKCS1 support, but didn't.  We did support encrypted PEM.
+
+It should be noted that Keyvault stores things in PKCS12 and PEM. Keyvault is not straight forward
+in how it works. Frankly, I'm in serious doubt that a regular Go user can get certs out of
+Keyvault's Go API.  
+
+Before I began working on MSAL I was re-writing the Keyvault Go API.  https://github.com/element-of-surprise/keyvault . It does the right things to extract cers for TLS now. 
+I was still working on the Cert() API and hadn't exposed the public surface when I stopped.
+
+Since we have representation from the Go SDK team, we might have them go bridge this problem in
+the current implementation using some of that code so its possible for our users to store the
+cert in Keyvault.
+
+## Notes:
+
+Do we need: AcquireTokenSilent()??  Seems like we could just bake this into other acquire calls automatically????
```

Comment:
@sangonzal  recently had a [proposal](https://identitydivision.visualstudio.com/DevEx/_git/AuthLibrariesApiReview/pullrequest/1948?_a=files&path=%2FConfidentialClient%2FClientCredentialCacheBehavior.md) to move confidential client flows to actually do the acquire token silent calls within the library instead of asking the developers to exclusively do it. So we really do not need it in the confidential client application acquire token calls where there is no user account associated. However, for public clients, acquire token silent call is still needed as acquire token silent would go in and find an access token for an account in the token cache that is associated with a user account. Developers would be the ones choosing which user accounts they want to do the cache look up for. That would look something like this : https://github.com/AzureAD/microsoft-authentication-library-for-python/blob/dev/sample/device_flow_sample.py#L49-L57

Reply:
Can you write an item that proposes the changes we need to do here?

Reply:
We are already tracking it here : https://github.com/AzureAD/microsoft-authentication-library-for-go/issues/48